### PR TITLE
Spill join

### PR DESCRIPTION
**Pull request by pnowojski**
First two commits are not part of this pull request (dependencies from other pull requests)

Last commit disables tests, that shows a bug from spill-v2 branch (previously not visible because of different session properties used in this test)

LookJoinOperator needs refactoring.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/spiller/SingleStreamSpiller.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+    /**
+     * Initiate spilling of single page. Returns completed future once spilling has finished.
+     */
+    CompletableFuture<?> spill(Page page);
+
+    /**
+     * Returns list of previously spilled Pages as a single stream.
+     */
+    Iterator<Page> getSpilledPages();
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/spiller/SingleStreamSpiller.java]: shouldn't it be called `readSpilledPages`?

File: presto-main/src/main/java/com/facebook/presto/spiller/SingleStreamSpiller.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import java.util.concurrent.CompletableFuture;
+
+@ThreadSafe
+public interface SingleStreamSpiller
+        extends Closeable
+{
+    /**
+     * Initiate spilling of pages stream. Returns completed future once spilling has finished.
+     */
+    CompletableFuture<?> spill(Iterator<Page> page);
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/spiller/SingleStreamSpiller.java:31]: I think this could have a default implementation which could use `spill(Page)` internally

File: presto-main/src/main/java/com/facebook/presto/spiller/BinarySpillerFactory.java [hunk 3]
```diff
@@ -20,6 +20,7 @@
 import com.google.common.util.concurrent.ListeningExecutorService;
 import com.google.common.util.concurrent.MoreExecutors;
 import com.google.inject.Inject;
+import sun.reflect.generics.reflectiveObjects.NotImplementedException;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/spiller/BinarySpillerFactory.java]: I think there is such exception in java.lang.

See UnsupportedOperationException.

File: presto-main/src/main/java/com/facebook/presto/spiller/PartitioningSpiller.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+public interface PartitioningSpiller
+        extends Closeable
+{
+    /**
+     * Partition page and enqueue partitioned pages to spill writers. Return completed future when finished.
+     */
+    PartitioningSpillResult partitionAndSpill(Page page);
+
+    /**
+     * Returns list of previously spilled Pages.
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/spiller/PartitioningSpiller.java]: mention about partition parameter

File: presto-main/src/main/java/com/facebook/presto/spiller/PartitioningSpiller.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+import java.util.concurrent.CompletableFuture;
+
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public interface PartitioningSpiller
+        extends Closeable
+{
+    /**
+     * Partition page and enqueue partitioned pages to spill writers. Return completed future when finished.
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/spiller/PartitioningSpiller.java]: update javadoc or drop

File: presto-main/src/main/java/com/facebook/presto/spiller/SpillerFactory.java [hunk 6]
```diff
@@ -14,13 +14,26 @@
...
+    default Spiller create(List<Type> types)
+    {
+        throw new UnsupportedOperationException();
+    }
 
-    SingleStreamSpiller createSingleStreamSpiller(List<Type> types);
+    default SingleStreamSpiller createSingleStreamSpiller(List<Type> types)
+    {
+        throw new UnsupportedOperationException();
+    }
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/spiller/SpillerFactory.java]: make this change to happen in the commit which introduced this method

File: presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+            PagesSerde.writePages(blockEncodingSerde, output, pageIterator);
+        }
+    }
+
+    @Override
+    public synchronized Iterator<Page> getSpilledPages()
+    {
+        waitForSpillsToFinish();
+        checkState(scheduledSpills.isEmpty(), "Can not read if writes have not finished");
+        return readPages(targetFile);
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java]: do not pass field as parameter to a method of the same class

File: presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+    }
+
+    private synchronized Iterator<Page> readPages(Path spillPath)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java]: if this method is synchronized only single thread can spill at a time with the same spiller, is that desired?

File: presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+@ThreadSafe
+public class BinaryFileSingleStreamSpiller
+        implements SingleStreamSpiller
+{
+    private final Path targetFile;
+    private final BlockEncodingSerde blockEncodingSerde;
+    private final SliceOutput output;
+    private final ListeningExecutorService executor;
+    private final Queue<CompletableFuture<?>> scheduledSpills = new ArrayDeque<>();
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java]: Instead of `Queue` you could a single `Future` and https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#allOf-java.util.concurrent.CompletableFuture...-

File: presto-main/src/main/java/com/facebook/presto/spiller/SingleStreamSpiller.java (hunk 1 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/spiller/SingleStreamSpiller.java]: I saw in the implementation that the returned `Pages` are in the same order in which they got spilled. Is this needed property for anything? If so can you mention about this in javadoc?

File: presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    private void writePage(Page page)
+    {
+        synchronized (output) {
+            PagesSerde.writePages(blockEncodingSerde, output, page);
+        }
+    }
+
+    private void writePages(Iterator<Page> pageIterator)
+    {
+        synchronized (output) {
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java]: why do you sometimes synchronize on `this` and sometimes on `output`? To me it looks like a bug, and there might be a race conditions between read and write operations.

File: presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+package com.facebook.presto.spi.block;
+
+import io.airlift.slice.Slice;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static java.util.stream.Collectors.toList;
+
+public class MaskedBlock
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java]: why it was added to SPI?

File: presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+    private int mapPosition(int position)
+    {
+        checkReadablePosition(position);
+        return positionsMapping[position + mappingOffset];
+    }
+
+    private void checkReadablePosition(int position)
+    {
+        if (position < 0 || position >= getPositionCount()) {
+            throw new IllegalArgumentException("position is not valid");
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java]: maybe `position is masked`?

File: presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static java.util.stream.Collectors.toList;
+
+public class MaskedBlock
+        implements Block
+{
+    private Block block;
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java]: final?

File: presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+    private int[] positionsMapping;
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java]: final

File: presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java (hunk 11 above)
Comment (kokosing) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java]: some javadoc would be nice. The name is wrong as it does not mask positions, but maps. I would name it `PositionMappingBlock`

File: presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+
+    public MaskedBlock(Block block, int[] positionsMapping, int mappingOffset, int mappingLength)
+    {
+        this.block = block;
+        this.positionsMapping = positionsMapping;
+        this.mappingOffset = mappingOffset;
+        this.mappingLength = mappingLength;
+    }
+
+    public Block getUnderlyingBlock()
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java]: It looks hacky

File: presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+
+    private int mapPosition(int position)
+    {
+        checkReadablePosition(position);
+        return positionsMapping[position + mappingOffset];
+    }
+
+    private void checkReadablePosition(int position)
+    {
+        if (position < 0 || position >= getPositionCount()) {
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java]: I would inline and use`mappingLength`

File: presto-main/src/main/java/com/facebook/presto/spiller/BinarySpillerFactory.java [hunk 17]
```diff
@@ -68,10 +66,4 @@ public SingleStreamSpiller createSingleStreamSpiller(List<Type> types)
     {
         return new BinaryFileSingleStreamSpiller(blockEncodingSerde, executor, spillPath);
     }
-
-    @Override
-    public PartitioningSpiller createPartitioningSpiller(List<Type> types, HashGenerator hashGenerator, int partitionsCount)
-    {
-        throw new NotImplementedException();
-    }
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/spiller/BinarySpillerFactory.java]: I think this belongs to different commit

File: presto-main/src/main/java/com/facebook/presto/spiller/GenericPartitioningSpiller.java [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+
+        for (int partition = 0; partition < partitionsCount; partition++) {
+            this.ignorePartitions[partition] = ignorePartitions.contains(partition);
+            pageBuilders[partition] = new PageBuilder(types);
+            spillers[partition] = spillerFactory.createSingleStreamSpiller(types);
+        }
+    }
+
+    @Override
+    public synchronized Iterator<Page> getSpilledPages(int partition)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/spiller/GenericPartitioningSpiller.java:70]: `readSpilledPages` ?

File: presto-main/src/main/java/com/facebook/presto/spiller/GenericPartitioningSpiller.java [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+        getFutureValue(flush(partition));
+        return spillers[partition].getSpilledPages();
+    }
+
+    @Override
+    public synchronized PartitioningSpillResult partitionAndSpill(Page page)
+    {
+        checkState(!readingStarted);
+        IntArrayList unspilledPositions = partitionPage(page);
+        CompletableFuture<?> future = flush();
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/spiller/GenericPartitioningSpiller.java:83]: flushing per each `Page` might be not a good idea. It is causing creation of small Pages. Maybe `Page` should be flushed only when it reaches some size threshold?

File: presto-main/src/test/java/com/facebook/presto/spiller/GenericPartitioningSpillerTest.java [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+            }
+        }
+    }
+
+    private static class PositiveAndNegativePartitionsHashGenerator
+            extends LocalPartitionGenerator
+    {
+        private final int bigintChannel;
+
+        public PositiveAndNegativePartitionsHashGenerator(int bigintChannel)
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/spiller/GenericPartitioningSpillerTest.java]: did you mean hashChannel?

File: presto-main/src/main/java/com/facebook/presto/spiller/GenericPartitioningSpiller.java [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+    private final PageBuilder[] pageBuilders;
+    private final LocalPartitionGenerator partitionGenerator;
+    private final SingleStreamSpiller[] spillers;
+    private boolean readingStarted;
+
+    public GenericPartitioningSpiller(
+            List<Type> types,
+            LocalPartitionGenerator partitionGenerator,
+            int partitionsCount,
+            Set<Integer> ignorePartitions,
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/spiller/GenericPartitioningSpiller.java]: when partition is ignored? when it has be spilled or not? I would rename it. The name is confusing.

File: presto-main/src/test/java/com/facebook/presto/spiller/GenericPartitioningSpillerTest.java [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+    private static final List<Type> TYPES = ImmutableList.of(BIGINT, VARCHAR, DOUBLE, BIGINT);
+    private static final LocalPartitionGenerator POSITIVE_AND_NEGATIVE_PARTITIONS_HASH_GENERATOR = new PositiveAndNegativePartitionsHashGenerator(0);
+    private final BlockEncodingSerde blockEncodingSerde = new BlockEncodingManager(new TypeRegistry(ImmutableSet.of(BIGINT, DOUBLE, VARBINARY)));
+    private final BinarySpillerFactory factory = new BinarySpillerFactory(blockEncodingSerde, new FeaturesConfig());
+
+    @Test
+    public void testFileSpiller()
+            throws Exception
+    {
+        try (PartitioningSpiller spiller = factory.createPartitioningSpiller(TYPES, POSITIVE_AND_NEGATIVE_PARTITIONS_HASH_GENERATOR, PARTITIONS_COUNT, ImmutableSet.of())) {
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/spiller/GenericPartitioningSpillerTest.java]: do not spill all partitions

File: presto-main/src/test/java/com/facebook/presto/spiller/GenericPartitioningSpillerTest.java [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+                ImmutableList.of(firstSpill, secondSpill),
+                ImmutableList.of(positiveRows, negativeRows));
+    }
+
+    private void testSpiller(List<Type> types, PartitioningSpiller spiller, List<List<Page>> spills, List<List<Page>> expectedPartitions)
+            throws ExecutionException, InterruptedException
+    {
+        for (List<Page> spill : spills) {
+            for (Page page : spill) {
+                spiller.partitionAndSpill(page).isBlocked().get();
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/spiller/GenericPartitioningSpillerTest.java]: check the unspilled positions

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java [hunk 24]
```diff
@@ -25,6 +25,8 @@
 import com.facebook.presto.spi.Page;
 import com.facebook.presto.spi.block.Block;
 import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spiller.SingleStreamSpiller;
+import com.facebook.presto.spiller.SpillerFactory;
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java:28]: I think that changes in this class belongs to different commit

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java [hunk 25]
```diff
@@ -82,26 +82,40 @@ public void tearDown()
...
-                {true, false, true},
-                {true, false, false},
-                {false, true, true},
-                {false, true, false},
-                {false, false, true},
-                {false, false, false}};
+                {true, true},
+                {true, false},
+                {false, true},
+                {false,  false}};
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java]: I would extract changes to this provider as separate commit

Comment (kokosing) [reply presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java]: is it used at all?

File: presto-tests/src/test/java/com/facebook/presto/tests/TestLocalBinarySpilledQueries.java [hunk 26]
```diff
@@ -311,4 +311,12 @@ public void testRightJoinWithNullValues()
     {
         super.testRightJoinWithNullValues();
     }
+
+    @Test(enabled = false)
```
Comment (kokosing) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestLocalBinarySpilledQueries.java]: can you add a link, or drop

File: presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java [hunk 27]
```diff
@@ -80,19 +108,50 @@ public OperatorContext getOperatorContext()
...
     {
-        return types;
+        return allTypes;
     }
 
     @Override
     public void finish()
     {
+        if (!finishing && lookupSourceSupplier.hasSpilled()) {
+            // Last LookupJoinOperator will handle spilled pages
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java]: instead of commenting, extract lines up to `count > 0` as method

File: presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java [hunk 28]
```diff
@@ -80,19 +108,50 @@ public OperatorContext getOperatorContext()
...
+            int count = lookupJoinsCount.decrementAndGet();
+            checkState(count >= 0);
+            if (count > 0) {
+                spiller = Optional.empty();
+            }
+            else {
+                // last LookupJoinOperator might not spilled at all, but other parallel joins could
+                // so we have to load spiller if this operator hasn't used one yet
+                ensureSpillerLoaded();
+                unspillNextLookupSource();
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java:126]: no need to unspill here, you can simply set spilledProbePages to an empty iterator and unspill only in `getOutput`

File: presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java [hunk 29]
```diff
@@ -49,26 +64,39 @@
     private boolean finishing;
     private long joinPosition = -1;
 
+    private Optional<PartitioningSpiller> spiller = Optional.empty();
+    private ListenableFuture<? extends LookupSource> spilledLookupSourceFuture = Futures.immediateFuture(null);
+    private ListenableFuture<?> spillInProgress = NOT_BLOCKED;
+    private Optional<Iterator<Integer>> spilledPartitions = Optional.empty();
+    private Optional<Iterator<Page>> spilledProbePages = Optional.empty();
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java]: I would rename it it as `unspilledProbePages`

Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java]: why they are optionals? I would just have empty iterators?

File: presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java [hunk 30]
```diff
@@ -16,31 +16,46 @@
 import com.facebook.presto.operator.LookupJoinOperators.JoinType;
 import com.facebook.presto.spi.Page;
 import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.block.Block;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java:19]: how about having a specialized `SpillableLookupJoinOperator` which could use `LookupJoinOperator` internally?

File: presto-main/src/main/java/com/facebook/presto/operator/ParallelHashBuildOperator.java [hunk 31]
```diff
@@ -193,10 +197,18 @@ public void finish()
         }
         finishing = true;
 
-        LookupSource lookupSource = index.createLookupSource(hashChannels, preComputedHashChannel, filterFunction);
-        lookupSourceSupplier.setLookupSource(partitionIndex, lookupSource);
+        if (spiller.isPresent()) {
+            lookupSourceSupplier.lookupSourceSpilled(partitionIndex, spiller.get());
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/ParallelHashBuildOperator.java]: rename the method to `setSpilledLookupSource`

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceSupplier.java [hunk 32]
```diff
@@ -79,28 +99,47 @@ public PartitionedLookupSourceSupplier(List<Type> types, List<Integer> hashChann
...
-            }
+        if (partitionedLookupSource != null) {
+            lookupSourceFuture.set(partitionedLookupSource);
+        }
+    }
+
+    public synchronized void lookupSourceSpilled(int partitionIndex, SingleStreamSpiller lookupSourceSpiller)
+    {
+        PartitionedLookupSource partitionedLookupSource;
+        synchronized (this) {
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceSupplier.java]: it is already synchronized

File: presto-tests/src/test/java/com/facebook/presto/tests/TestLocalBinarySpilledQueries.java [hunk 33]
```diff
@@ -102,4 +103,212 @@ public void testRowFieldAccessorInAggregate()
     {
         super.testRowFieldAccessorInAggregate();
     }
+
+    @Test(expectedExceptions = RuntimeException.class, expectedExceptionsMessageRegExp = "Spilling for RIGHT and FULL outer joins is unsupported")
```
Comment (kokosing) [inline presto-tests/src/test/java/com/facebook/presto/tests/TestLocalBinarySpilledQueries.java]: generally I don't like `expectedException` to be used in that way. See `com.facebook.presto.tests.AbstractTestQueryFramework#assertQueryFails`. With similar method you could test all the possibilities in one method and it would much more readable.

**Review by kokosing — COMMENTED** [review]
not bad ;)

File: presto-main/src/main/java/com/facebook/presto/spiller/SingleStreamSpiller.java (hunk 2 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/spiller/SingleStreamSpiller.java:31]: dropped as discussed (because of `Iterator<Page>` materialization and sequential disk writes)

File: presto-main/src/main/java/com/facebook/presto/spiller/SingleStreamSpiller.java (hunk 1 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/spiller/SingleStreamSpiller.java]: it doesn't have to be "disk" based spill. It could be compressed memory or anything else, so `getSpilledPages` is a little bit more general.

File: presto-main/src/main/java/com/facebook/presto/spiller/BinarySpillerFactory.java (hunk 3 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/spiller/BinarySpillerFactory.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/spiller/PartitioningSpiller.java (hunk 5 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/spiller/PartitioningSpiller.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/spiller/PartitioningSpiller.java (hunk 4 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/spiller/PartitioningSpiller.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/spiller/SpillerFactory.java (hunk 6 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/spiller/SpillerFactory.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java (hunk 9 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java]: this list changes dynamically

File: presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java (hunk 10 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java]: it's not a bug. Those two methods are executed in executor and are long blocking operations (writes do disk). Added some comments

File: presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java (hunk 7 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java (hunk 8 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpiller.java]: it is desired, but that means that reading blocks writes. It is possible to call `spill` methods from multiple threads/operators.

File: presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java (hunk 11 above)
Comment (pnowojski) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java]: because all other blocks are in SPI? They have to be there?

Renamed and added some doc.

File: presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java (hunk 13 above)
Comment (pnowojski) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java]: fixed

File: presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java (hunk 14 above)
Comment (pnowojski) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java]: fixed

File: presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java (hunk 15 above)
Comment (pnowojski) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java]: This method was not used actually.

File: presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java (hunk 16 above)
Comment (pnowojski) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java]: this one is more generic

File: presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java (hunk 12 above)
Comment (pnowojski) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/MaskedBlock.java]: this throws also for other cases, like negative position or position way to big (larger then masked block)

File: presto-main/src/main/java/com/facebook/presto/spiller/BinarySpillerFactory.java (hunk 17 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/spiller/BinarySpillerFactory.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/spiller/GenericPartitioningSpiller.java (hunk 21 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/spiller/GenericPartitioningSpiller.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/spiller/GenericPartitioningSpiller.java (hunk 18 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/spiller/GenericPartitioningSpiller.java:70]: as before, dropped

File: presto-main/src/main/java/com/facebook/presto/spiller/GenericPartitioningSpiller.java (hunk 19 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/spiller/GenericPartitioningSpiller.java:83]: it already works as you want. Pages are only flushed when their PageBuilder `isFull`

File: presto-main/src/test/java/com/facebook/presto/spiller/GenericPartitioningSpillerTest.java (hunk 22 above)
Comment (pnowojski) [reply presto-main/src/test/java/com/facebook/presto/spiller/GenericPartitioningSpillerTest.java]: done

File: presto-main/src/test/java/com/facebook/presto/spiller/GenericPartitioningSpillerTest.java (hunk 23 above)
Comment (pnowojski) [reply presto-main/src/test/java/com/facebook/presto/spiller/GenericPartitioningSpillerTest.java]: done

File: presto-main/src/test/java/com/facebook/presto/spiller/GenericPartitioningSpillerTest.java (hunk 20 above)
Comment (pnowojski) [reply presto-main/src/test/java/com/facebook/presto/spiller/GenericPartitioningSpillerTest.java]: Renamed to `hashChannel` but actually it has to be bigint, maybe `bigintHashChannel` would be better name?

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java (hunk 24 above)
Comment (pnowojski) [reply presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java:28]: They belong here. I have renamed commit to `Pass spiller parameters to ParallelHashBuildOperator`

File: presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java (hunk 25 above)
Comment (pnowojski) [reply presto-main/src/test/java/com/facebook/presto/operator/TestHashJoinOperator.java]: Extracted. Yes, in `testMemoryLimit`

File: presto-main/src/main/java/com/facebook/presto/operator/ParallelHashBuildOperator.java (hunk 31 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/operator/ParallelHashBuildOperator.java]: fixed

File: presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceSupplier.java (hunk 32 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceSupplier.java]: dropped method synchronization

File: presto-tests/src/test/java/com/facebook/presto/tests/TestLocalBinarySpilledQueries.java (hunk 33 above)
Comment (pnowojski) [reply presto-tests/src/test/java/com/facebook/presto/tests/TestLocalBinarySpilledQueries.java]: but those tests are overridden :/ I can not do that without coping test cases

File: presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java (hunk 30 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java:19]: please review rewritten commit (and refactor commit which is at begin of the PR)

File: presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java (hunk 28 above)
Comment (pnowojski) [reply presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java:126]: please review rewritten commit (and refactor commit which is at begin of the PR)

**Review by pnowojski — COMMENTED** [review]
please review rewritten commit (and refactor commit which is at begin of the PR)

File: presto-main/src/test/java/com/facebook/presto/spiller/GenericPartitioningSpillerTest.java (hunk 20 above)
Comment (kokosing) [reply presto-main/src/test/java/com/facebook/presto/spiller/GenericPartitioningSpillerTest.java]: ok
